### PR TITLE
Runnable pipeline on test data

### PR DESCRIPTION
**Pull request by mumichae**
Create a test dataset for development.
Testing can be done directly in the repository if `config/test_data.yaml` is used without modification. This PR doesn't address testing in pytest (`tests/`) or setting up a CI, which will be addressed in different PRs.

### CODE REVIEW

File: envs/env_vars_activate.sh [hunk 1]
```diff
@@ -11,3 +11,7 @@ export LDFLAGS="`gsl-config --libs` ${LDFLAGS_OLD}"
 R_LIBS_OLD=$R_LIBS
 export R_LIBS_OLD
 export R_LIBS=""
+
+LD_LIBRARY_PATH_OLD=$LD_LIBRARY_PATH
+export LD_LIBRARY_PATH
```
Comment (LuckyMD) [inline envs/env_vars_activate.sh:16]: missing an `_OLD` here

File: configs/test_data.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  test_data:
+    batch_key: batch
+    label_key: celltype
+    organism: mouse
+    assay: expression
+    file: data/adata_norm.h5ad
+
+r_env : scIB-R-integration
+py_env : scIB-python
+conv_env : scIB-R-conversion
```
Comment (LuckyMD) [inline configs/test_data.yaml:13]: isn't scIB-python meant to be the conv_env?

File: Snakefile [hunk 3]
```diff
@@ -1,15 +1,16 @@
-from config.ScibConfig import *
+from ScibConfig import *
 
 #configfile: "config.yaml"
 cfg = ParsedConfig(config)
 wildcard_constraints:
-    hvg = "hvg|full_feature"
+    hvg="hvg|full_feature",
+    scaling="scaled|unscaled",
+    method="[^/]+"
```
Comment (LuckyMD) [inline Snakefile:8]: what does the method statement mean here? Is this a regexp about the first letter not being a `/`?

File: README.md [hunk 4]
```diff
@@ -27,7 +27,25 @@ and `env_vars_deactivate.sh` to `etc/conda/deactivate.d/` to every environment.
...
+Furthermore, to ensure that `rpy2` uses the correct R version, the `LD_LIBRARY_PATH` variable has to be set accordingly.
+First, determine the path of the environment using `rpy2`
+```
+echo $CONDA_PREFIX
+```
+Then deactivate the environment to prevent unwanted effects.
+Next, set the `LD_LIBRARY_PATH` variable in the `${CONDA_PREFIX}/etc/conda/activate.d/activate-r-base.sh`
+```
+LD_LIBRARY_PATH_OLD=$LD_LIBRARY_PATH
+export LD_LIBRARY_PATH
```
Comment (LuckyMD) [inline README.md:39]: missing an `_OLD` here.

File: README.md [hunk 5]
```diff
@@ -27,7 +27,25 @@ and `env_vars_deactivate.sh` to `etc/conda/deactivate.d/` to every environment.
...
+export LD_LIBRARY_PATH="${CONDA_PREFIX}/lib/R/lib/"
+```
+and the following lines in `${CONDA_PREFIX}/etc/conda/deactivate.d/deactivate-r-base.sh`
+```
+LD_LIBRARY_PATH=LD_LIBRARY_PATH_OLD
+export LD_LIBRARY_PATH
+unset LD_LIBRARY_PATH_OLD
+```
```
Comment (LuckyMD) [inline README.md:47]: you have this in the `activate_env.sh` and `deactivate_env.sh` files already... why not just refer to those faileds that they should be moved to the relevant directories?

File: ScibConfig.py [hunk 6]
```diff
@@ -75,16 +75,18 @@ def get_from_method(self, method, key):
...
+    def get_hvg(self, wildcards, output_pattern, **kwargs):
         n_hvgs = self.get_feature_selection(wildcards.hvg)
         if n_hvgs == 0:
             return ""
         if self.get_from_method(wildcards.method, "R"):
-            path_parts = adata_path.split('.')
-            path_parts[-2] += '_hvg'
-            hvg_path = '.'.join(path_parts)
+            p = Path(expand(output_pattern, **wildcards, **kwargs)[0])
+            hvg_path = (p.parent / f'{p.stem}_hvg').with_suffix(p.suffix)
```
Comment (LuckyMD) [inline ScibConfig.py:89]: this is really cool... didn't know these operators were overloaded for Path variables.

File: Snakefile [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
 
 include: "scripts/preprocessing/Snakefile"
 include: "scripts/integration/Snakefile"
 include: "scripts/metrics/Snakefile"
 include: "scripts/visualization/Snakefile"
-include: "scripts/metrics/cc_score/Snakefile"
```
Comment (LuckyMD) [inline Snakefile:12]: was this something that is just not used anymore? I.e. a separate cc scoring rule you used for testing?

File: data/generate_data.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import scIB
+
+
+def get_adata_rand_batch(pca=False, n_top_genes=None, neighbors=False):
+    adata = sc.datasets.paul15()
+    adata.obs['celltype'] = adata.obs['paul15_clusters']
+    np.random.seed(42)
+    adata.obs['batch'] = np.random.randint(1, 5, adata.n_obs)
+    adata.obs['batch'] = adata.obs['batch'].astype(str)
+    adata.obs['batch'] = adata.obs['batch'].astype("category")
```
Comment (LuckyMD) [inline data/generate_data.py:12]: can just string together `astype(str).astype("category")`, no?

Comment (LuckyMD) [reply data/generate_data.py:12]: or just use category directly?

File: data/generate_data.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    sc.pp.pca(adata_ref)
+    sc.pp.neighbors(adata_ref)
+    sc.tl.umap(adata_ref)
+
+    # merge cell type labels
+    sc.tl.ingest(adata, adata_ref, obs='louvain')
+    adata_concat = adata_ref.concatenate(adata, batch_categories=['ref', 'new'])
+    adata_concat.obs.louvain = adata_concat.obs.louvain.astype('category')
+    # fix category ordering
+    adata_concat.obs.louvain.cat.reorder_categories(adata_ref.obs.louvain.cat.categories, inplace=True)
```
Comment (LuckyMD) [inline data/generate_data.py:46]: why does category order matter?

File: data/generate_data.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    from pathlib import Path
+
+    parser = argparse.ArgumentParser(description='Generate test dataset')
+    parser.add_argument('-f', '--force', action='store_true')
+    args = parser.parse_args()
+
+    filepath = Path(__file__).parent / 'adata_norm.h5ad'
+
+    if not filepath.exists() or args.force:
+        adata = get_adata_rand_batch()
```
Comment (LuckyMD) [inline data/generate_data.py:68]: never call the other function? Why not give options?

**Review by LuckyMD — CHANGES_REQUESTED** [review]
A few typos and a few clarifications. @danielStrobl can you check if you can get the test pipeline running in server environment?

File: README.md (hunk 5 above)
Comment (mumichae) [reply README.md:47]: These scripts didn't work for me, as `gsl-config` doesn't seem to be installed on my system. However only `LD` needed to be set in order to get `rpy2` running correctly. Also, Daniel mentioned something about `r-essentials` solving some of the issues

File: Snakefile (hunk 3 above)
Comment (mumichae) [reply Snakefile:8]: Yes, I had issues with `snakemake` not finding `method=R/seurat`

File: Snakefile (hunk 7 above)
Comment (mumichae) [reply Snakefile:12]: We decided to remove cc_score completely, as we didn't end up needing it

File: Snakefile (hunk 3 above)
Comment (LuckyMD) [reply Snakefile:8]: I guess that's an issue with the config file then, no? Not really sth we have hardcoded like this somehwere?

File: README.md [hunk 11]
```diff
@@ -47,16 +47,52 @@ unset LD_LIBRARY_PATH_OLD
...
+This is useful, in order to ensure that the installation was successful before moving on to a larger dataset.
+More information on how to use the data generation script can be found in `data/README.md`.
+
+### Setup Configuration File
+
+The parameters and input files are specified in config files, that can be found in `configs/`.
+In the `DATA_SCENARIOS` section you can define the input data per scenario.
+The main input per scenario is a preprocessed `.h5ad` file of an anndata with batch and cell type annotations.
+
+TODO: explain different entries
```
Comment (LuckyMD) [inline README.md:66]: just a reminder here :)

File: configs/test_iso_AWS.yaml [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    label_key: celltype
+    organism: mouse
+    assay: expression
+    file: data/adata_norm.h5ad
+  test_data_rep:
+    batch_key: batch
+    label_key: celltype
+    organism: mouse
+    assay: expression
+    file: data/adata_norm.h5ad
```
Comment (LuckyMD) [inline configs/test_iso_AWS.yaml:15]: Could we move DATA_SCENARIOS to the bottom please? Looks a bit better there imho.

**Review by LuckyMD — CHANGES_REQUESTED** [review]
Some small beautifications, but this is looking really good :).

File: Snakefile (hunk 3 above)
Comment (mumichae) [reply Snakefile:8]: We save R method output in a separate `R/` directory, which, when given the path, leads to ambiguity whether to use `method=seurat` from the R integration & conversion rules or `methd=R/seurat` from the python integration rule

Comment (LuckyMD) [reply Snakefile:8]: Ah, I see. Okay, thanks!

File: README.md (hunk 11 above)
Comment (mumichae) [reply README.md:66]: I'd opt to move these TODOs to a new PR and focus on other issues first

File: data/generate_data.py (hunk 8 above)
Comment (mumichae) [reply data/generate_data.py:20]: If I don't convert to str, conversion into R via `rpy2` will throw an error. But I agree, a one-liner is nicer

File: data/generate_data.py (hunk 9 above)
Comment (mumichae) [reply data/generate_data.py:58]: To properly combine the cluster labels of both datasets, I guess. I copied the code from the [scanpy tutorial](https://scanpy-tutorials.readthedocs.io/en/latest/integrating-data-using-ingest.html) and will add the link in the docstring

File: data/generate_data.py (hunk 10 above)
Comment (mumichae) [reply data/generate_data.py:80]: Yes, might want to consider that. However, the PBMC dataset did not make it through the pipeline, as it doesn't have enough gene features. I kept the code, in case we might want to use it, but would also consider removing it, if we don't need it

File: README.md (hunk 4 above)
Comment (mumichae) [reply README.md:39]: 957bc20

File: envs/env_vars_activate.sh (hunk 1 above)
Comment (mumichae) [reply envs/env_vars_activate.sh:16]: 957bc20

File: configs/test_iso_AWS.yaml (hunk 12 above)
Comment (mumichae) [reply configs/test_iso_AWS.yaml]: 6640afb

File: data/generate_data.py (hunk 9 above)
Comment (mumichae) [reply data/generate_data.py:63]: 375d957

File: data/generate_data.py (hunk 8 above)
Comment (mumichae) [reply data/generate_data.py]: 375d957

File: README.md (hunk 5 above)
Comment (mumichae) [reply README.md:47]: It might also be nicer to have all R-specific variables in the `${CONDA_PREFIX}/etc/conda/activate.d/activate-r-base.sh` and `${CONDA_PREFIX}/etc/conda/deactivate.d/deactivate-r-base.sh`, respectively

**Review by LuckyMD — COMMENTED** [review]
This looks pretty good. Did you just merge the new scIB-python env changes from master here?

File: data/generate_data.py (hunk 10 above)
Comment (LuckyMD) [reply data/generate_data.py:85]: sounds good.

File: README.md (hunk 5 above)
Comment (LuckyMD) [reply README.md:47]: Hmm... I think we can probably stick with env variables for now. If this doesn't work for you, could you just adapt the README to say there are 2 options to get this to work?

**Comment by LuckyMD** [comment]
@danielStrobl could you please test this so that we can merge?

**Comment by danielStrobl** [comment]
Hey! I've tested the pipeline with the new envs and I'm still getting some errors:
Metrics scanorama:
```
ValueError: batch "3" of batch_key "batch" has unequal number of entries before and after integration.before: 564 after: 550
```
Metrics with precomputed embeddings:
```
  File "scripts/metrics/metrics.py", line 249, in <module>
    trajectory_=trajectory_
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 1855, in metrics
    cluster=False, n=n_isolated, verbose=False) if silhouette_ else np.nan
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 325, in isolated_labels
    score = score_isolated_label(adata, label_key, cluster_key, label, cluster=cluster, verbose=verbose)
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 391, in score_isolated_label
    score = silhouette(adata_tmp, group_key='group', **kwargs)
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 50, in silhouette
    raise KeyError(f'{embed} not in obsm')
KeyError: 'X_pca not in obsm'
```
The integration part seems to work fine

**Comment by mumichae** [comment]
Thanks for testing the setup.

I'm getting the same error above with scanorama and mnn. This happens only when I add a batch effect to the counts in the `paul_15` dataset. Haven't figured out why this happens.

For the embedding error, use the `isolated_label_ASW_embed` branch of `scib`. You can find the PR here: https://github.com/theislab/scib/pull/213

**Comment by mumichae** [comment]
I identified the cause of errors in `mnnpy` and `scanorama`, see the [PR in scib](https://github.com/theislab/scib/pull/215).
TL;DR `Anndata.concatenate` renames annotations of the `batch_key`, if that key exists. In our test data, the batch key is `batch`, which is also the default for `batch_key` in `Anndata.concatenate`

**Comment by mumichae** [comment]
@danielStrobl Feel free to test the pipeline with the https://github.com/theislab/scib/tree/fix_concatenate for integration and `master` for the metrics. You can use the rule `integration` with the former branch until all methods run through, then switch to master and run the rest of the pipeline.

```commandline
snakemake --config-file config/test_data.yaml -j10 integration
```

**Comment by LuckyMD** [comment]
The fix_concatenate PR is good enough to be merged after a small confirmation on your part. Then this will be easier to test here again as well :). @danielStrobl it would be great if you could retest this as soon as @mumichae gives the go ahead with the other fix!

**Comment by mumichae** [comment]
scib PR is merged. I pushed some bugfixes, so don't forget to update @danielStrobl

**Review by LuckyMD — DISMISSED** [review]
Approved if this runs for @danielStrobl

**Comment by danielStrobl** [comment]
I've tested with the new envs and the updates now, everything is running through without any error

**Review by danielStrobl — APPROVED** [review]

**Comment by mumichae** [comment]
@LuckyMD Is this ready to be merged? Solving merge conflicts has revoked the approval. Also, I have adapted the documentation for environment variables, including the `QT_QPA_PLATFORM='offscreen'` variable as mentioned [here](https://github.com/theislab/scib/issues/209)

**Review by LuckyMD — APPROVED** [review]
in general this is good... just 1 comment on the README

File: README.md [hunk 13]
```diff
@@ -22,23 +22,82 @@ To install the envs, use
...
+### Setting Environment Parameters
+Some parameters need to be added manually to the conda environment in order for packages to work correctly.
+For example, all environments using R need `LD_LIBRARY_PATH` set to the conda R library path.
+If that variable is not set, `rpy2` might reference the library path of a different R installation that might be on your system.
 
-To load the config file run `snakemake --configfile config.yaml`.
-Define the number of CPU threads you want to use with `snakemake --cores N_CORES`. To produce an overview of tasks that will be run, use `snakemake -n`.
-To run the pipeline, simply run `snakemake`.
+Environment variables are provided in `env_vars_activate.sh` and `env_vars_deactivate.sh` and should be copied to `etc/conda/activate.d/env_vars.sh`
+and `etc/conda/deactivate.d/env_vars.sh` respectively.
```
[the remark this answers is not in the document]
Comment (mumichae) [reply README.md:36]: Added some explanation on how and where to copy these files

File: README.md (hunk 5 above)
Comment (mumichae) [reply README.md:98]: Moved back to using the env files only and added more info on them in the README

**Comment by mumichae** [comment]
Would be good to have this merged asap, so I can move on to the scanpy update etc. here

File: ScibConfig.py [hunk 14]
```diff
@@ -75,16 +63,33 @@ def get_from_method(self, method, key):
...
+            wildcards: snakemake.io.Wildcards,
+            output_pattern: str = None,
+            **kwargs
+    ) -> str:
+        """
+        Get hvg parameter for integration run scripts
+        :param wildcards: wildcards passed by Snakemake containing at least 'hvg' key
+        :param output_pattern: file pattern with placeholders 'hvg'
+            Only needed for R integration methods to get path of separate HVG file
+            minimal example: 'output_dir/{hvg}.h5ad'
```
Comment (LuckyMD) [inline ScibConfig.py:82]: I think we only pass a separate HVG file to Seurat, but not other R methods. Have you changed this to do so for all R methods?

File: envs/env_vars_deactivate.sh [hunk 15]
```diff
@@ -8,6 +8,11 @@ LDFLAGS=$LDFLAGS_OLD
 export LDFLAGS
 unset LDFLAGS_OLD
 
-R_LIBS=$R_LIBS_OLD
-export $R_LIBS
-unset R_LIBS_OLD
+# Unset rpy2 library path
+LD_LIBRARY_PATH=LD_LIBRARY_PATH_OLD
```
Comment (LuckyMD) [inline envs/env_vars_deactivate.sh:12]: Missing a `$` here in front of `LD_LIBRARY_PATH_OLD`

File: sanity_check/metrics_compare.R [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    x = old_label,
+    y = new_label
+  ) +
+  theme_classic() +
+  theme(legend.position = 'bottom')
+
+ggsave(
+  file.path('data', 'sanity', paste(comparison, 'png', sep = '.')),
+  width = 8, height = 9
+)
```
Comment (LuckyMD) [inline sanity_check/metrics_compare.R:67]: This file should probably not be added here

File: scripts/integration/Snakefile [hunk 17]
```diff
@@ -16,52 +15,52 @@ rule integration_run_python:
         command: {params.cmd}
         hvgs: {params.hvgs}
         cell type option: {params.cell_type}
+        output: {output[0]}
         """
     params:
         batch_key = lambda wildcards: cfg.get_from_scenario(wildcards.scenario, key="batch_key"),
         cell_type = cfg.get_celltype_option_for_integration,
-        hvgs      = lambda wildcards, input: cfg.get_hvg(wildcards, input.adata[0]),
+        hvgs      = cfg.get_hvg,
```
Comment (LuckyMD) [inline scripts/integration/Snakefile:23]: you don't need an input argument here?

File: scripts/metrics/metrics.py [hunk 18]
```diff
@@ -136,6 +136,7 @@
     if (type_ == "embed"):
         n_hvgs = None
         embed = "X_emb"
+        precompute_pca = False
```
Comment (LuckyMD) [inline scripts/metrics/metrics.py:139]: this is from iso label fix? I guess you merged master into here?

**Review by LuckyMD — CHANGES_REQUESTED** [review]
At least 1 typo (missing $) and a few questions on how you now store hvg files and some missing input args. Also 1 file added that doesn't belong in this PR (visualization of iso label fix results). Looks very close though.

File: sanity_check/metrics_compare.R (hunk 16 above)
Comment (mumichae) [reply sanity_check/metrics_compare.R:67]: I needed it to compare different dependencies for the pipeline. Would you prefer it in the reproducibility repo instead?

File: scripts/integration/Snakefile (hunk 17 above)
Comment (mumichae) [reply scripts/integration/Snakefile:23]: no, Snakemake is smart :D 

This notation works when the only input is `wildcards`. I removed the input file pattern, as the python script doesn't require any separate HVG input (it's computed in the script).

File: scripts/metrics/metrics.py (hunk 18 above)
Comment (mumichae) [reply scripts/metrics/metrics.py:139]: Yes, I could remove that to reduce runtime slightly

Comment (mumichae) [reply scripts/metrics/metrics.py:139]: It's not merged from scib master though, I added it here for testing

File: ScibConfig.py (hunk 14 above)
Comment (mumichae) [reply ScibConfig.py:82]: In the `integration_run_r` rule (check `scripts/integration/Snakefile`) I always pass a HVG file to the script. I'm assuming that script uses the HVGs for all R methods in question

Comment (mumichae) [reply ScibConfig.py:82]: From what I can see, the count object is subset by HVGs for all R methods, if it is given as an option

Comment (LuckyMD) [reply ScibConfig.py:82]: Hah... you're right... the subsetting is done later for R methods... only Seurat passes the list of HVGs. This is fine... was worried for a sec ^^.

File: sanity_check/metrics_compare.R (hunk 16 above)
Comment (LuckyMD) [reply sanity_check/metrics_compare.R:67]: Hmm... this is a script we are using for our checks and not something that we show or expect others to use, no?  Would you want to expose this to the public?

File: scripts/metrics/metrics.py (hunk 18 above)
Comment (LuckyMD) [reply scripts/metrics/metrics.py:139]: This is part of the iso label fix though, no? I mean... I think it's correct to do this. But not initially the remit of this PR. So just wondering if you also have this in scIB at the moment...

File: scripts/integration/Snakefile (hunk 17 above)
Comment (LuckyMD) [reply scripts/integration/Snakefile:23]: cool :). Okay!

File: scripts/metrics/metrics.py (hunk 18 above)
Comment (mumichae) [reply scripts/metrics/metrics.py:139]: I just double-checked and the results don't seem to change if I precompute PCA. Will remove it then

Comment (LuckyMD) [reply scripts/metrics/metrics.py]: this is what caused us to miss the iso label issue (as there was a PCA fallback). So I would keep this in here. I just wanted to make sure whether this code was already in scIB as well (before file moving).

**Review by LuckyMD — APPROVED** [review]
All good, just please add the precompute PCA false again (and check this is consistent across repos), then we can merge.

File: scripts/metrics/metrics.py (hunk 18 above)
Comment (mumichae) [reply scripts/metrics/metrics.py]: The variable is not set to False for embed input in scib master

Comment (LuckyMD) [reply scripts/metrics/metrics.py]: okay, then nvm. mergeable :)
